File: articles/iteration-experimentation-design-thinking/go/experiment_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Experiment struct {
	Name                string
	LearningGain        float64
	UpdateFlexibility   float64
	ExpectedImprovement float64
	ResidualRisk        float64
	Value               float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func experimentValue(experiment Experiment) float64 {
	return 0.35*experiment.LearningGain +
		0.25*experiment.UpdateFlexibility +
		0.25*experiment.ExpectedImprovement -
		0.15*experiment.ResidualRisk
}

func loadExperiments(path string) []Experiment {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var experiments []Experiment

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 5 {
			log.Fatalf("malformed row: %v", row)
		}

		experiment := Experiment{
			Name:                row[0],
			LearningGain:        parseFloat(row[1]),
			UpdateFlexibility:   parseFloat(row[2]),
			ExpectedImprovement: parseFloat(row[3]),
			ResidualRisk:        parseFloat(row[4]),
		}
		experiment.Value = experimentValue(experiment)
		experiments = append(experiments, experiment)
	}

	return experiments
}

func main() {
	path := "../data/raw/experiments_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	experiments := loadExperiments(path)

	sort.Slice(experiments, func(i, j int) bool {
		return experiments[i].Value > experiments[j].Value
	})

	fmt.Println("rank,experiment,experiment_value")
	for i, experiment := range experiments {
		fmt.Printf("%d,%s,%.6f\n", i+1, experiment.Name, experiment.Value)
	}
}

File: articles/iteration-experimentation-design-thinking/cpp/experiment_value_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct Experiment {
    std::string name;
    double learning_gain;
    double update_flexibility;
    double expected_improvement;
    double residual_risk;
};

struct ScoredExperiment {
    std::string name;
    double value;
};

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;

    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }

    return fields;
}

std::vector<Experiment> read_experiments(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<Experiment> experiments;
    std::string line;

    std::getline(file, line); // header

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 5) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        experiments.push_back(
            Experiment{
                fields[0],
                to_double(fields[1]),
                to_double(fields[2]),
                to_double(fields[3]),
                to_double(fields[4])
            }
        );
    }

    return experiments;
}

double experiment_value(const Experiment& experiment) {
    constexpr double wl = 0.35;
    constexpr double wu = 0.25;
    constexpr double we = 0.25;
    constexpr double wr = 0.15;

    return wl * experiment.learning_gain
         + wu * experiment.update_flexibility
         + we * experiment.expected_improvement
         - wr * experiment.residual_risk;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/experiments_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto experiments = read_experiments(input_path);
        std::vector<ScoredExperiment> scored;

        for (const auto& experiment : experiments) {
            scored.push_back(ScoredExperiment{experiment.name, experiment_value(experiment)});
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredExperiment& a, const ScoredExperiment& b) {
                return a.value > b.value;
            }
        );

        std::cout << "rank,experiment,experiment_value\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].value << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
